hpx/util/detail/basic_function.hpp

```diff
@@ -200,117 +249,120 @@ namespace hpx { namespace util { namespace detail
+        using vtable = typename std::conditional<
+                Copyable,
+                detail::function_vtable<R(Ts...)>,
+                detail::unique_function_vtable<R(Ts...)>
+            >::type;
+        using serializable_vtable = serializable_function_vtable<vtable>;
+        using base_type = function_base<R(Ts...), Copyable>;
 
     public:
-        typedef R result_type;
-
         basic_function() noexcept
           : base_type()
+          , serializable_vptr(nullptr)
         {}
 
-        basic_function(basic_function&& other) noexcept
-          : base_type(static_cast<base_type&&>(other))
-        {}
+        template <typename F>
+        void assign(F&& f)
+        {
+            using target_type = typename std::decay<F>::type;
 
-        basic_function& operator=(basic_function&& other) noexcept
+            base_type::assign(std::forward<F>(f));
+            if (!base_type::empty())
+            {
+                serializable_vptr = get_serializable_vtable<target_type>();
+            }
+        }
+
+        void swap(basic_function& f) noexcept
         {
-            base_type::operator=(static_cast<base_type&&>(other));
-            return *this;
+            base_type::swap(f);
+            std::swap(serializable_vptr, f.serializable_vptr);
         }
 
     private:
         friend class hpx::serialization::access;
 
-        void load(serialization::input_archive& ar, const unsigned version)
+        void save(serialization::output_archive& ar, unsigned const version) const
         {
-            this->reset();
-
-            bool is_empty = false;
-            ar >> is_empty;
+            bool const is_empty = base_type::empty();
+            ar << is_empty;
             if (!is_empty)
             {
-                std::string name;
-                ar >> name;
+                std::string const name = serializable_vptr->name;
+                ar << name;
 
-                this->vptr = detail::get_vtable<vtable>(name);
-                this->object = this->vptr->load_object(
-                    this->storage, function_storage_size, ar, version);
+                serializable_vptr->save_object(object, ar, version);
             }
         }
 
-        void save(serialization::output_archive& ar, const unsigned version) const
+        void load(serialization::input_archive& ar, unsigned const version)
         {
-            bool is_empty = this->empty();
-            ar << is_empty;
+            base_type::reset();
+
+            bool is_empty = false;
+            ar >> is_empty;
             if (!is_empty)
             {
-                std::string function_name = this->vptr->name;
-                ar << function_name;
+                std::string name;
+                ar >> name;
+                serializable_vptr = detail::get_serializable_vtable<vtable>(name);
 
-                this->vptr->save_object(this->object, ar, version);
+                vptr = serializable_vptr->vptr;
```

Comment:
Is there a way to avoid the extra storage for `vptr` inside `serializable_vptr`? It seems superfluous to me.

Reply:
Maybe, I get lost around the function registration logic.. How would it work?

Reply:
maybe by adding a `serializable_vtable::get_vptr` virtual function, which gives us back the needed vtable pointer?

Reply:
No, the serializable vtable type only knows the function vtable type, not the target callable type. That gets erased in the constructor, where the function is registered. I think the function vtable pointer would have to be stored alongside the serializable vtable pointer in the intrusive registration factory, where it could later be retrieved from by function name, but I find that code hard to navigate.

Reply:
Ok, that might be something for a different PR then.